### ISSUE TITLE
Support lensfun database

### ISSUE THREAD

**Issue by AdrianEddy** [gyroflow/gyroflow#150]
https://lensfun.github.io/

**Comment by Entropy512** [comment]
I'm working on trying to understand your distortion handling code - it looks like it's pretty hardcoded for the OpenCV fisheye model right now.

I see in https://github.com/gyroflow/gyroflow/blob/master/src/core/gpu/opencl_undistort.cl#L46 that it looks like you merged OpenCV's remap and the model to execute in a single kernel, vs OpenCV's original implementation of generating a map from the model, then using a generic remap kernel.  The original OpenCV method is pretty similar to how darktable also does it - it uses lensfun to generate the map, then its own OpenCL kernels to perform the image transformations.  (lens_distort_bicubic for example in https://github.com/darktable-org/darktable/blob/master/data/kernels/basic.cl#L1079, along with https://github.com/darktable-org/darktable/blob/master/src/iop/lens.cc#L815 to generate the mapping using lensfun's ApplySubpixelGeometryDistortion function ( https://lensfun.github.io/manual/latest/structlfModifier.html#a217fef568888472fbc3d472794f5bb38 ).  It looks like OpenCV only supports nearest neighbor and linear in the OpenCL kernel, and you're using the (bi)linear interpolation implementation?

Also it looks like you're using different distortion parameters for each line in the image as part of the rolling shutter correction algorithm - https://github.com/gyroflow/gyroflow/blob/master/src/core/gpu/opencl_undistort.cl#L64 - where are these parameters generated?  I haven't yet managed to trace through and find/understand the other components of the RS correction mechanism.  (Hopefully maybe tonight I'll find it...)

Is there a reason you merged the model generation and remap into one function?  (I'm guessing that is related to the RS corrector, and am wondering if there's some way they can be decoupled...)

**Comment by AdrianEddy** [comment]
The undistortion params are generated in `src/core/undistortion/frame_transform.rs:at_timestamp`: https://github.com/gyroflow/gyroflow/blob/master/src/core/undistortion/frame_transform.rs#L74

If rolling shutter correction is not enabled, this will produce only 1 3x3 matrix. If it's enabled it will produce *height* 3x3 matrices, one for each row of the image. Camera matrix is multiplied with the rotation matrix here (in Rust on CPU), and GPU kernels take that precomputed matrix and distortion coefficients to calculate final mapping.

Yes, we're using (bi)linear interpolation, but I think that's the cause of #179 so I want to change it to lanczos4 interpolation (maybe leave linear for preview)

The reason to merge it like that and do in single OpenCL kernel was for performance. `initUndistortRectifyMap` on CPU and then uploading the maps to GPU was really slow. We can split it into multiple OpenCL functions though, that should give us more flexibility. This code was actually the very first code written for new Gyroflow 4 months ago, there's definitely room for improvement.

Do you have experience with that kind of stuff and would you like to help organize it better to support more interpolation methods and lens models?

**Comment by Entropy512** [comment]
I'm a bit rusty with OpenCL, last time I did anything with it was around 4 years ago attempting to make some improvements/changes to darktable.  I no longer use dt (loooong story there...).  My big weak point is I have zero experience with rust, but I can try to maybe poke at this at least to figure out a way to segregate the map functions.

Shouldn't it be possible to upload the main distortion map only once, leaving you only to need to send the rolling shutter rotation matrices and actual image data on a per-frame basis?  Darktable is probably not a good example here of such a technique since it doesn't process consecutive video frames, as a result, they recompute the map for every image processed instead of precomputing it somewhere and uploading it once.  DT does at least have examples of applying a map using lanczos and bicubic interpolation.  Hmm...  Would setting the lens distortion map in new() and not reset it in undistort_image() of https://github.com/gyroflow/gyroflow/blob/master/src/core/gpu/opencl.rs work?  Do kernel args only get uploaded when set, or do they get re-uploaded on every enqueue?

As far as #179 - wouldn't the bilinear interpolation in undistort also affect OFX???

FYI, ffmpeg has some rather neat tricks for getting conditionals for configuration parameters outside of OpenCL kernels - they literally build the source file and compile it when the filter initializes.  https://github.com/FFmpeg/FFmpeg/blob/master/libavfilter/vf_tonemap_opencl.c#L129 - not sure if there's benefit to getting that aggressive at nuking conditionals from kernels here.

**Comment by AdrianEddy** [comment]
> As far as #179 - wouldn't the bilinear interpolation in undistort also affect OFX???

It would, that's why I'm puzzled. But I figured lanczos4 should provide better results than bilinear so we should still do it.
OFX works on 32-bit floats though (input and output pixel buffers), but we do interpolation with 32-bit floats there regardless of the pixel input, so it should be the same even with 8-bit input pixels

> Would setting the lens distortion map in new() and not reset it in undistort_image() of https://github.com/gyroflow/gyroflow/blob/master/src/core/gpu/opencl.rs work? Do kernel args only get uploaded when set, or do they get re-uploaded on every enqueue?

I don't know, maybe it's possible, but what exactly do you want to upload once? We still need to upload 3x3 matrices per every frame, so what are we saving here? This would require to rewrite the entire math there though. Current kernels themselves are **really** fast, I mean, they run under 1ms for 4k frame, there's no need to optimize it further. Uploading pixels to GPU and downloading from GPU is the bottleneck here, but it's very hard to avoid that if we want to have portability across all different graphics APIs + hw video decoders and encoders.

**Comment by Entropy512** [comment]
Performance wise, probably not much compared to the current architecture.  Organization wise - right now since the distortion model is being implemented inside the kernel instead of by creating a map that is uploaded once, it's really hard to use alternative distortion models.  Having conditionals for the distortion model inside the kernel would be a bit messy, and also the distortion model would have to be implemented in CPU, WGL, OpenCL, and GLSL.

If the kernel could have the distortion map uploaded once, then each distortion model (OpenCV fisheye, PTLens, Sony's model, possibly also ACM models?) could then be implemented solely in CPU, precalculated, and hopefully uploaded once.  You'd still have to upload the rolling shutter rotation matrices, but at least that's only height*9, as opposed to width*height*2 (the distortion map dimensions, with an (xs,ys) pair for each output (x,y) position)

As to OFX
I recall the OFX instructions mentioning loading the OFX node in Fusion and not the main Resolve node graph - is there a reason for that?  (I have not really played with anything OFX in Resolve yet, or Fusion at all).  If I recall correctly, Resolve heavily prefers to linearize everything before feeding it to Fusion - so the OFX plugin may be operating on linear data, which might wind up behaving differently when interpolating.  (In fact good luck doing vignetting or TCA corrections without linear data...)

**Comment by AdrianEddy** [comment]
> Performance wise, probably not much compared to the current architecture. Organization wise - right now since the distortion model is being implemented inside the kernel instead of by creating a map that is uploaded once, it's really hard to use alternative distortion models. Having conditionals for the distortion model inside the kernel would be a bit messy, and also the distortion model would have to be implemented in CPU, WGL, OpenCL, and GLSL.

That would be great to separate the distortion model from the rest of the kernel, but I don't have enough knowledge about distortion to implement that. Contribution is more than welcome and I'll be happy to assist on any technical Rust/GPU stuff if you can provide math for it.

> I recall the OFX instructions mentioning loading the OFX node in Fusion and not the main Resolve node graph - is there a reason for that? (I have not really played with anything OFX in Resolve yet, or Fusion at all). If I recall correctly, Resolve heavily prefers to linearize everything before feeding it to Fusion - so the OFX plugin may be operating on linear data, which might wind up behaving differently when interpolating. (In fact good luck doing vignetting or TCA corrections without linear data...)

I didn't investigate that issue yet so I'm not sure what's going on there. I need to do it before final v1.0.0 release though, I'll let you know when I have some more insight

**Comment by Entropy512** [comment]
Hopefully in addition to having more time this weekend to poke at actual camera testing (vs weekdays I frequently have more opportunities to dig through code but no camera access...), I'll be able to maybe poke at Resolve to see if perhaps it's a linear vs. nonlinear sort of thing.  I do recall seeing some comparisons regarding scaling algorithms that indicated that doing image scaling was best done with linearized data, but plenty of software gets away with "OK" results scaling gamma-compressed image data, distortion correction may be similar.  Doing linearization for best quality is something that's probably best left to OFX since one of Resolve's primary functions is converting various nonlinear luminance encoding schemes to/from each other (it's a major part of color management), implementing to/from linear transforms is a wonderful can of worms that you don't want to open if you don't have to.

As far as where the distortion functions line, one of your commits actually serves as pretty good documentation of what is part of the distortion algorithm:

https://github.com/gyroflow/gyroflow/commit/0cb1db223ef0529a908fdf45f4243b9f36e98352

What I see there is:
Calculate rotated x/y based on the first rotation matrix (which you corrected to the middle one in a later commit)
if _w<0 (I'm still trying to wrap my head around what _w is...  Looks like some sort of scaling/zooming factor?):
Determine the source x and source y for the "post-rotation" x and y by taking its radius from center and applying the distortion model

That _w scaling factor is going to be problematic, since I'm assuming it changes per frame at a minimum (FOV scaling???), unless perhaps it can be implemented as a modifier into a precomputed lookup table, something like (x,y) -> (wx, wy) -> (sx, sy), where the (wx,wy)->(sx,sy) conversion is handled.  But now you'd be doing two bilinear lookups - one to scale (x,y) by the w factor, and then to do a lookup of the scaled result into the distortion map.  (basically, all of the math you have after calculating r then calculating theta from r is the distortion mapping function, which would have to be replaced with a different function of r for other distortion models.)

I'll try to find a better way to write/describe this...  But the key to implementing most other distortion models is going to be to implement alternate formulas that transform "r" to "scale" in that area, potentially doing some unit conversions for r.

**Comment by Entropy512** [comment]
As far as a possible explanation for 179:

http://www.ericbrasseur.org/gamma.html?i=1

Solving that properly can be...  fun...  :(

**Comment by AdrianEddy** [comment]
Thanks, that's a very interesting read. We have this task: #12 and wgpu textures support a [lot of formats](https://docs.rs/wgpu/latest/wgpu/enum.TextureFormat.html), and they can convert srgb->linear in shader->srgb back in buffer so I'm very curious if it can give us a quality improvement, just by implementing that one (which is fairly simple to do)

**Comment by AdrianEddy** [comment]
> As far as where the distortion functions line, one of your commits actually serves as pretty good documentation of what is part of the distortion algorithm:

The current distortion model also includes camera matrix, and that's computed outside of kernel in `frame_transform.rs`, so the pure rotation matrix (from quaternion) is multiplied with the fisheye camera matrix and the invese of it is the `undistortion_params` in the kernel. 
So the distortion model is split to two different places and that's problematic here. I'm pretty sure it can be improved, but I don't know how to approach that

**Comment by Entropy512** [comment]
Getting more familiar with OpenCV has always been on my TODO list, time to do more fiddling.  :)

I think translating different coordinate systems between distortion models would not be too problematic (even deriving a synthetic camera matrix from known parameters - e.g. if you know focal length and physical sensor width, then you can get very close to a camera matrix for a video that is known to be full sensor width:
cx = w/2
cy = h/2
fx = focallen*imagewidth/sensorwidth
fy = fx for most lenses (anamorphic would be the exception here)

For example, for https://github.com/gyroflow/gyroflow/blob/master/resources/camera_presets/Sony/Sony_A7S3_Tamron%2028-200%40100mm_NO-EIS_4k_16by9_3840x2160-59.94fps.json#L50

The camera matrix you derived is VERY close to:
cx = 3840/2 = 1920
cy = 2160/2 = 1080
fx = 100*3840/36 = 10666.66

The one thing that really can't be translated reliably between camera and distortion models is (sx,sy) = f(x,y) for distortion itself.

For the OpenCV fisheye model, f(x,y) is doing:
r = sqrt(x^2+y^2)
theta = atan(r)
etc etc

There's no way, for example, to translate to/from OpenCV's Brown-Conrady model thanks to that pesky theta=atan(r), or to/from Sony's spline-function-of-r model

As a side note, as far as OpenCV's fisheye vs pinhole models - the paper that describes calibrating the fisheye model says it should work for non-fisheye lenses but they don't really provide testing/results for doing so.  Seeing huge high-order factors such as https://github.com/gyroflow/gyroflow/blob/master/resources/camera_presets/Sony/Sony_A7S3_Tamron%2028-200%40100mm_NO-EIS_4k_16by9_3840x2160-59.94fps.json#L50 makes me think the distortion parameters are starting to be used to fight the model itself here.

Gotta go, will poke at this later.

**Comment by Entropy512** [comment]
Just leaving this as a reference for anyone else who might be following this and wanting to get a better feel for the coordinate transforms:

https://learnopencv.com/geometry-of-image-formation/ is one of the better documents I've seen describing the behavior of the camera matrix

**Comment by Entropy512** [comment]
It probably would be good to document theory of operation somewhere.  As a start, perhaps you and @ElvinC can confirm my rough understanding of what's happening here:

From https://learnopencv.com/geometry-of-image-formation/ - assume world coordinate system is X,Y,Z, and camera-transformed homgenous coordinates is u',v',w', with camera-transformed non-homogenous coordinates as u,v - homogenous coordinates for camera-transformed allows preserving depth/distance information via w.

In frame_transform, a "virtual camera" is created derived from the original camera matrix and FOV scaling.  Unlike the "real" camera, this one is using the rectilinear/pinhole model and not the fisheye model.  Its parameters are stored in new_k

When rolling shutter correction is active, a "virtual camera" transform R that is the opposite of the real camera's rotation is calculated for each real-camera video row.  This transform is then multiplied by new_k.  The pseudoinverse of new_k*R is taken to get a matrix that will transform camera-homogenous coordinates u',v',w' of the virtual camera to real-camera's X,Y,Z coordinate system

Now in opencl_undistort, the process is as follows:
Instead of using u,v for non-homogenous coordinates in the output video, x,y are used - perhaps these should be renamed for consistency with OpenCV nomenclature/documentation?  From here on, I'll reference them as u,v for my sanity
For each pixel u,v, that pixel is projected into camera world coordinates by assuming u' = u, v' = v, and w=1 and multiplying those by the inverted camera transform matrix
The results of this are X,Y,Z world coordinates, but named _x, _y, _w - perhaps these should be renamed too?

These X,Y,Z world coordinates are then transformed into real-camera image coordinates as follows:
**NOTE:  Model-specific math that can't be translated between models starts here**
r = sqrt((X/Z)^2+(Y/Z)^2) - the division by Z is why https://github.com/opencv/opencv/pull/6801 was created
theta = atan(r)
derive distortion adjustment from k1-k4 and theta^2 etc
determine non-homogenous real-camera image coordinates u,v directly, since there's no need for depth/distance here
gyroflow actually calls this variable uv - but then assigns the first element of it to sy (maybe rename to su? - I understand the variable.x and variable.y nomenclature for float2 is probably hardcoded but the rest can be renamed)
**Model-specific math ends here**
source v is chosen (named sy in the implementation), new transform is determined based on source v

Reprojection is repeated with the new transform

Observation:  I think that if the camera is pitching up/down, when you reproject with a new transform, sy may change significantly.  It may be necessary to iterate this multiple times until it converges...  At least with some initial experiments with an A7M4 + FE85 (26.8ms readout time instead of the 8.7 of the A7S3 you tested with), I'm seeing good horizontal/rotation performance but there's still significant vertical shake/jello - will upload an example and file a separate issue on this later tonight or tomorrow.

I think it would be beneficial to update variable names for consistency - I can take a first crack at a pull request to do that for at least the opencl stuff tomorrow, and if you like the changes, I'll do them against the other implementations.

Observation 2:  It looks like you're not using accelerometer data for linear translations right now?  For cameras with subject distance information available (Sony at least, and possibly could be a user parameter for other cameras), you could probably implement this by translating the virtual camera and choosing w' appropriately to put the world-coordinate image plane at the appropriate distance.  Probably won't be useful for drones, but might be beneficial for handheld cameras at closer subject distances.

**Comment by AdrianEddy** [comment]
@Entropy512 From my understanding everything you said here is correct, yes. 
I agree to the variable rename, currently it's a mix of variables from python, opencv camera stuff and `remap`'s variable names (which is not related to the camera model). It'll be good to document it and make consistent, we can also refactor out interpolation to a separate function at this point, just the clean the sources a bit

> Reprojection is repeated with the new transform

The main reprojection happens at this second iteration here, the first one is calculated only when rolling shutter correction is active and is used to determine the row in source pixels (*sy*) to use correct rotation matrix for this row. At the beginning I just used output row here (*y*), without this second calculation but obviously that was a shortcut and it didn't work for extreme rolling shutter cases

> Observation 2: It looks like you're not using accelerometer data for linear translations right now? 

That's correct, and I agree it would be nice to add support for that as well

Thank you for taking time to poke with this and your interest in this area and feel free to create any PRs 🙂

Perhaps we should rename this issue to general "Lens model refactoring"

**Comment by Entropy512** [comment]
Either that or create a separate issue for general lens model refactoring that ties to all of the lens model issues.

I'll start poking at doing some variable name clarifications and adding some comments later today or tomorrow.  BTW is there a good way to determine which code path was used?  It LOOKS like OpenCL takes precedence over WGPU but I can't be sure.  I am guessing my machine is taking the OCL path but other than intentionally breaking it and looking for an effect, I'm not sure.

**Comment by AdrianEddy** [comment]
if you see `[INFO] OpenCL Platform: ...` in console log, and you don't see big `OpenCL ERROR`, then OpenCL is used, it's always preferred if it works.

If you see `[DEBUG] WGPU adapter: ....` then wgpu was used

otherwise it uses CPU and you can tell by looking at your task manager CPU usage (all cores maxed out basically) :)

**Comment by Entropy512** [comment]
OK, based on what little I saw in the log, I was guessing OpenCL but that confirms it.

**Comment by AdrianEddy** [comment]
@Entropy512 With latest commit, I refactored the kernels and made it easy to implement and add alternative lens models
check out [src/core/stabilization/distortion_models](https://github.com/gyroflow/gyroflow/tree/master/src/core/stabilization/distortion_models) and [updated kernels code](https://github.com/gyroflow/gyroflow/blob/master/src/core/gpu/opencl_undistort.cl)

**Comment by Entropy512** [comment]
Thanks, I'll try and take a look later this week.  Sorry I've been absent for a while as far as working on this, I need to get back to poking at this now that the weather is nicer and I'll be going out with the camera a lot more.  :)

**Comment by AdrianEddy** [comment]
Some update to clarify the issue:
- The lens models from LensFun are already implemented here: https://github.com/gyroflow/gyroflow/tree/master/src/core/stabilization/distortion_models
- The missing part is figure out how to rescale coefficients and take crop factor into account.
- These lens models were never used so they also need to be tested. 
- All the lensfun code is available in their repository, so it should be easy to understand how to use it and debug any issues
- I Imagine we can add the lensfun repository as a submodule to https://github.com/gyroflow/lens_profiles  and implement parsing of their XMLs in lens_profile_database.rs

**Comment by samuel-asleep** [comment]
@AdrianEddy is this issue still open for me to work on or has it already been addressed ?

**Comment by yasumorishima** [comment]
/attempt #150

Plan:
- Use the `lensfun` crate (pure-Rust port, verified equivalent to upstream within 4.9e-4 pixels) to read the LensFun XML database, rather than vendoring XMLs via a submodule.
- Add a `load_from_lensfun(maker, model, lens_name)` path to `lens_profile_database.rs`.
- Map lensfun's `ptlens`/`poly3`/`poly5` parameters into gyroflow's existing `distortion_models` (same math, per the 2023-10-20 note).
- Handle coefficient rescaling + crop_factor via lensfun's `Modifier` API.
- Add unit tests comparing gyroflow's existing distortion_models output against lensfun-rs's reference `apply_geometry_distortion` output.

Feasibility verified in my fork: `gyroflow-core` builds clean, `lensfun 0.7` adds cleanly (only `roxmltree` as extra transitive dep), `Database::load_bundled()` + lens lookup + `Modifier::new()` all work against the bundled DB.

**Comment by aimqwest** [comment]
/attempt
AIMQWEST submission: implementing Lensfun database support incrementally with parsing + integration tests.

**Comment by bazx-bit** [comment]
Hi @AdrianEddy, is this bounty still active?

I’m interested in working on Lensfun database support. Before starting, I’d like to confirm the expected direction:

1. Parse/import Lensfun camera and lens profiles into Gyroflow’s existing lens profile format.
2. Map Lensfun calibration fields to the current Gyroflow distortion/FOV model where compatible.
3. Add a small UI/import path or CLI conversion path, depending on what fits the project better.
4. Include test profiles and documentation for supported Lensfun fields and unsupported cases.

If this is still wanted, can you confirm the preferred integration approach and bounty details?

**Comment by jamilahmadzai** [comment]
/attempt #150

Plan:
- Add bundled Lensfun database import support for supported Poly3, Poly5, and PTLens distortion entries.
- Convert Lensfun coefficients with crop factor, aspect ratio, real focal length, and distortion-center handling into Gyroflow lens profiles.
- Register generated Lensfun entries in the existing lens profile database/search flow, then regenerate the selected profile for the current video dimensions when loaded.
- Cover the conversion with tests, including a pixel-level geometry comparison against Lensfun's own Modifier output.

**Comment by teddychan-dev** [comment]
/attempt #150

Implemented Lensfun database import in gyroflow-core, following the approach outlined in [the clarification](https://github.com/gyroflow/gyroflow/issues/150#issuecomment-1772624387):

- Parse Lensfun XMLs (`poly3`/`poly5`/`ptlens`) dropped into the lens profiles directory
- Port of `lfLens::InterpolateDistortion` (Hermite spline over focal length, crop-factor based calibration set selection)
- Coefficient rescaling via the existing (previously unused) `rescale_coeffs` functions, taking the calibration sensor's crop factor and aspect ratio into account — i.e. the TODO block at the bottom of `poly3.rs`
- Camera matrix from real focal length + pixel pitch of the target sensor

Validated numerically against Lensfun git master (`lfModifier::ApplyGeometryDistortion` on a controlled database): matches to < 0.005 px, including Hermite-interpolated focals and a full-frame calibration used on a 1.5 crop camera. PR incoming shortly.

**Comment by BabaVictim** [comment]
Algora still lists this as a $200 open bounty, but I can see several active Lensfun implementations (#1141, #1169, #1170, #1174, #1182, and #1186). Before duplicating work: is the bounty still active, which existing direction is preferred, and is there a concrete acceptance gap that remains? If a focused gap-fix would qualify independently, I can take that once the expected behavior is confirmed.

**Comment by AdrianEddy** [comment]
please don't create any more PRs for this, There's already way too many of them

**Comment by ParsingMAINFRAME** [comment]
/attempt #150

Implementation plan (following @AdrianEddy's stated architecture from the issue thread):

1. Add the lensfun database as a submodule to `gyroflow/lens_profiles` and implement XML parsing in `lens_profile_database.rs` — both `<lens>` and `<camera>` entries, with mount-based camera↔lens matching and per-camera crop factors.
2. Coefficient rescaling: compute `real_focal` per model (ptlens: `focal·(1−a−b−c)`, poly3: `focal·(1−k1)`, poly5: `focal`), then hugin-scale (`hypot(36,24)/crop/hypot(aspect,1)/2`) and apply the existing `rescale_coeffs()` in `distortion_models/`. Synthesize the pinhole camera matrix from real focal + sensor diagonal so resolution scaling stays in `frame_transform.rs` as today.
3. Validation: numeric tests comparing undistorted points against lensfun's own `Modifier` output (reference implementation), plus crop-mismatch tolerance rule (calibration sensor ≥ target, 0.96 tolerance) ported from mod-coord.cpp.
4. Wire the lensfun profiles into the existing UI search (`prepare_list_for_ui`) with correct camera_brand/camera_model vs lens fields, and hook per-frame focal-length interpolation (`InterpolateDistortion` Hermite spline) for zoom lenses using the existing `get_interpolated_lens_at` plumbing.

Two questions before I invest the effort:
- Do you still want the lensfun DB vendored as a submodule in `gyroflow/lens_profiles` (with a companion PR there), or would you prefer a bundled snapshot in this repo?
- Are you available to review in the coming weeks? Happy to keep the PR scoped and iterative.

**Comment by younes-bkb** [comment]
/attempt #150

Plan:
1. Survey existing open LensFun-related PRs and current main lens-profile code paths.
2. Implement LensFun database import/mapping into Gyroflow's lens profile format with auto-update where it fits existing patterns.
3. Add tests / verification notes and open a focused PR with `/claim #150`.

**Comment by stffinfcti** [comment]
/attempt #150

Implementation plan (per @AdrianEddy's clarification comment):
1. Parse Lensfun XMLs (`<camera>` + `<lens>` entries, multi-mount, multi-focal `<calibration>` blocks) from a `lensfun/` dir in the lens profiles path — designed so lensfun's `data/db` can be vendored as a submodule of `gyroflow/lens_profiles`.
2. Coefficient rescaling + crop factor via the existing `rescale_coeffs` in `stabilization/distortion_models` (poly3/poly5/ptlens) — port of lensfun's hugin coordinate scaling and `real_focal` derivation.
3. Camera↔lens mount matching + crop-factor compatibility rule (port of `lfLens` calibration-set selection).
4. Zoom-lens interpolation through the existing `interpolations` profile mechanism.
5. Unit tests in `src/core` verifying coefficient conversion against lensfun's own formulas (these models were never used/tested).

**Comment by stffinfcti** [comment]
@AdrianEddy Noted on not creating more PRs for this — I won't open one. For reference, my attempt implements the lensfun XML db in gyroflow-core (parser for cameras/mounts/multi-calibration lenses, coefficient rescaling via the existing distortion models, lazy per-camera profile generation rather than materializing the db) on the stffinfcti/gyroflow bounty/150-lensfun-db branch; it would also need a small gyroflow/lens_profiles companion to vendor the XMLs. If a fresh PR is still welcome I'll open it — otherwise the branch stays put.
